File: Medidor.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <chrono>
using namespace std;

// Criba tradicional para obtener primos base
vector<int> cribaBase(int limite)
{
    vector<bool> esPrimo((size_t)limite + 1, true);
    vector<int> primos;
    esPrimo[0] = esPrimo[1] = false;

    for (int i = 2; i * i <= limite; i++)
    {
        if (esPrimo[i])
        {
            for (int j = i * i; j <= limite; j += i)
                esPrimo[j] = false;
        }
    }

    for (int i = 2; i <= limite; i++)
        if (esPrimo[i])
            primos.push_back(i);

    return primos;
}

// Criba segmentada
long long contarPrimos(long long n)
{
    long long limite = sqrt(n);
    vector<int> primos = cribaBase((int)limite);

    long long inicio = limite;
    long long fin = 2 * limite;

    long long total = primos.size();

    while (inicio < n)
    {
        if (fin > n)
            fin = n;

        vector<bool> esPrimo((size_t)(fin - inicio + 1), true);

        for (int p : primos)
        {
            long long base = (inicio / p) * p;
            if (base < inicio)
                base += p;
            for (long long j = base; j <= fin; j += p)
                esPrimo[(size_t)(j - inicio)] = false;
        }

        for (long long i = inicio; i <= fin; i++)
            if (esPrimo[(size_t)(i - inicio)])
                total++;

        inicio = fin + 1;
        fin = inicio + limite;
    }

    return total;
}

int main()
{
    long long n = 20000000; // 20 millones
    auto inicio = chrono::high_resolution_clock::now();

    long long total = contarPrimos(n);

    auto fin = chrono::high_resolution_clock::now();
    chrono::duration<double> tiempo = fin - inicio;

    cout << "Total de primos <= " << n << ": " << total << endl;
    cout << "Tiempo: " << tiempo.count() << " segundos\n";

    return 0;
}

File: Medidor.rs
use std::time::Instant;

fn criba_base(limite: usize) -> Vec<usize> {
    let mut es_primo = vec![true; limite + 1];
    let mut primos = vec![];

    es_primo[0] = false;
    es_primo[1] = false;

    for i in 2..=((limite as f64).sqrt() as usize) {
        if es_primo[i] {
            let mut j = i * i;
            while j <= limite {
                es_primo[j] = false;
                j += i;
            }
        }
    }

    for i in 2..=limite {
        if es_primo[i] {
            primos.push(i);
        }
    }

    primos
}

fn contar_primos(n: usize) -> usize {
    let limite = (n as f64).sqrt() as usize;
    let primos_base = criba_base(limite);

    let mut total = primos_base.len();
    let mut inicio = limite;
    let mut fin = inicio + limite;

    while inicio < n {
        if fin > n {
            fin = n;
        }

        let tamaño = fin - inicio + 1;
        let mut es_primo = vec![true; tamaño];

        for &p in &primos_base {
            let mut base = (inicio / p) * p;
            if base < inicio {
                base += p;
            }

            let mut j = base;
            while j <= fin {
                es_primo[j - inicio] = false;
                j += p;
            }
        }

        for i in inicio..=fin {
            if es_primo[i - inicio] {
                total += 1;
            }
        }

        inicio = fin + 1;
        fin = inicio + limite;
    }

    total
}

fn main() {
    let n: usize = 20_000_000;

    let inicio = Instant::now();
    let total = contar_primos(n);
    let duracion = inicio.elapsed();

    println!("Total de primos <= {}: {}", n, total);
    println!("Tiempo: {:.4} segundos", duracion.as_secs_f64());
}
